### PR TITLE
Implement disable default mounts via command line

### LINKED ISSUE #22364: Allow disabling default mounts via `podman machine init` command
### Feature request description

There are predefined mounts for machine init. It is possible to override them with `-v "/mnt:/mnt"` series of arguments, but it is not currently possible to disable them, when running `podman machine init` command. Later on it is possible to disable them by editing machine config json file.

This covers edge cases, when one wants to run podman machine as isolated from host as possible or cases, where due to some specifics it is not possible to use mounts from host.

### Suggest potential solution

Add special treatment to the `-v ""` option (empty string value), which will override the defaults, but not actually add anything to the list.

Could be implemented with a patch as little as this (not including documentation).

```diff
diff --git a/pkg/machine/shim/volume.go b/pkg/machine/shim/volume.go
index a145f5847..e67e577db 100644
--- a/pkg/machine/shim/volume.go
+++ b/pkg/machine/shim/volume.go
@@ -8,6 +8,9 @@ import (
 func CmdLineVolumesToMounts(volumes []string, volumeType vmconfigs.VolumeMountType) []*vmconfigs.Mount {
        mounts := []*vmconfigs.Mount{}
        for i, volume := range volumes {
+               if volume == "" {
+                       continue
+               }
                var mount vmconfigs.Mount
                tag, source, target, readOnly, _ := vmconfigs.SplitVolume(i, volume)
                switch volumeType {
```

### Have you considered any alternatives?

Manually editing machine config json after machine init.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by arixmkii**
Fixes #22364 

This PR adds specific handling of the command line arguments `-v ""` - special handling for empty string. `-v` being present suppresses defaults from the config, so, it is sufficient to implement it as no-op to achieve empty mounts. This happens as early as possible (before any machine specific logic would be called).

Additional change is to allow using QEMU w/o virtiofsd binary if there are no mounts needed and the binary will not be used to run additional daemon.

And the final change is to disable default mounts for e2e tests, which are not supposed to have any mounts. Implementation is a bit hacky it checks if the command is `machine init` and then traverses args to see if there are mounts requested, if no mounts in the list it will inject empty mount arguments and continue.

#### Does this PR introduce a user-facing change?

```release-note
Allow `podman machine init` w/o default mounts via command line arguments `-v ""`
```

### CODE REVIEW

**Comment by mheon** [comment]
I don't think the `-v ""` syntax will work. A new flag (`--no-default-mounts` or similar) would be more acceptable.

**Comment by arixmkii** [comment]
`-v ""` was tested on Windows host. So, from technical standpoint it works. It might still be not desirable from the API perspective standpoint.

**Comment by mheon** [comment]
That's mostly my concern - feels like poor user interface

**Comment by arixmkii** [comment]
I'm open to reimplement it, when the final decision is agreed within the team.

**Comment by mheon** [comment]
@Luap99 @ashley-cui @l0rd Thoughts on this one? I'm pretty firmly in favor of a separate flag but could be convinced otherwise

**Comment by rhatdan** [comment]
Since I can find no precedence for this, I guess I would go with new option.

But why not just edit containers.conf and remove the volumes?

**Comment by ashley-cui** [comment]
I'm on team edit containers.conf, but okay with a new flag.

**Comment by l0rd** [comment]
@arixmkii can you please provide more details on the use case? What's the use case that requires to disable the defaults volumes? And why editing the conf file doesn't work in this case?

**Comment by arixmkii** [comment]
Editing config would do the trick. I just expected there to be API to achieve the same effect, but there was no. Right now there is no sure way to init machine with mounts disabled w/o checking config in advanced. Example - running e2e machine tests would always use mounts in all tests with the default configuration and there is no way to force this behavior in tests only.

This actually could be considered as niche or minor issue. I created original issue about this and as this was not strictly rejected I submitted this follow up. If this is considered as not needed both this PR and the issue can be closed with "won't implement." In this case I can resubmit only a change for checking if mounts are requested before checking virtiofsd dependency.

@l0rd my scenario is the system, where I reinstall (often with prune everything uninstall or sometimes restoring that system from snapshot) Podman releases rebuilt from main head often and I can't have mounts operational. It's very specific and can be mitigated with some sort of automation on my side.

**Comment by Luap99** [comment]
I am fine with `-v ""` and actually would prefer it over a new option. I don't find the argument just edit the config very strong. You can have more than one machine and may not want mounts for only one so changiing th econfig over and over is not helpful to users.

File: pkg/machine/e2e/config_test.go
Comment (Luap99) [inline pkg/machine/e2e/config_test.go:204]: This does not look right. First of all this argument manipulation happens in the wrong place. It would need to happen for buildCmd() in the machineInit type.

Second I really really think tests must use the default volumes, it is critical that these mounts work. Most tests may not use them but it is still important that our code works and does not contain flakes etc.. (https://github.com/containers/podman/pull/23118) And yes this one test there actually depend on the default mounts.

**Review by Luap99 — CHANGES_REQUESTED** [review]

File: pkg/machine/e2e/config_test.go
Comment (arixmkii) [reply pkg/machine/e2e/config_test.go:204]: This part is indeed dirty and hacky. I need to get more familiar with the testing framework setup to improve this.

I noticed numerous tests, which were designed specifically for testing volume mounts, so, I didn't expect there will be some test, which implicitly test default ones. Probably the behavior from that test could be adjusted by loading the config as an app would do and manually converting it into chain of `-v "/some:/some" -v "/other:/other"`, but I'm not sure if it worth doing or if it is better to keep the things as is for testing as they have proved to be working fine already.

Alternatively I could keep everything as is and try to add dedicated tests for the new API option checking that it will create no mounts. And if at later point the adjustment of default testing setting would be needed it can be done in a separate PR. WDYT?

Comment (Luap99) [reply pkg/machine/e2e/config_test.go:204]: I find it important that we test defaults. Adding explicit machine mounts means we are no longer testing how most users use it.

**Comment by l0rd** [comment]
> @l0rd my scenario is the system, where I reinstall (often with prune everything uninstall or sometimes restoring that system from snapshot) Podman releases rebuilt from main head often and I can't have mounts operational. It's very specific and can be mitigated with some sort of automation on my side.

Thank you @arixmkii. Given the rarity of the use case we should avoid a new option. But this PR fixes the issue with an empty list of volume (`-v ""` creates a broken machine). So it makes sense to merge it as soon as you have addressed @Luap99 comments.

I have tested it on macOS with libkrun and it worked well 👍

File: pkg/machine/e2e/config_test.go
Comment (arixmkii) [reply pkg/machine/e2e/config_test.go]: Removed this workaround. Also, added a dedicated test for this command line.

**Comment by arixmkii** [comment]
macOS machine is legitimate. It seems that macOS mounts some additional paths for system purposes (I have no macOS hardware to check). I will disable new test on macOS.

**Comment by arixmkii** [comment]
Disabled test for both applehv and libkrun. Would appreciate if someone could share output of `findmnt -t virtiofs` from applehv and libkrun machines with default configuration to confirm my guess that there are additional active mounts to ones defined in standard config.

**Comment by Luap99** [comment]
@arixmkii Default macos mounts are defined here https://github.com/containers/common/blob/main/pkg/config/default_darwin.go#L16

**Comment by Luap99** [comment]
Ah I think the thing is that we mount rosetta payload via virtiofs on applehv thus you see the extra mount there.

**Comment by arixmkii** [comment]
@Luap99 Thank you! But, unfortunately, I don't need default ones, I need actual. I'm guessing that Rossetta binfmt is being mount as virtiofs from the host, but I can't prove it w/o testing. In this case there will be a virtiofs mount on macOS even, when no user mounts are set for VM.My current believe is that this is what lead to a test failure on macOS.

**Review by Luap99 — APPROVED** [review]
LGTM, I will put it on my list to fix the test to ensure they work with applehv/libkrun but that should not block the merge.

**Comment by rhatdan** [comment]
/lgtm
thanks @arixmkii

**Pull request merged**
